gym/core.py

```diff
@@ -244,6 +245,46 @@ def spec(self):
     def class_name(cls):
         return cls.__name__
 
+    @property
+    def action_space(self):
+        if self._action_space is None:
+            return self.env.action_space
+        return self._action_space
+
+    @action_space.setter
+    def action_space(self, space):
```

Comment:
I'm assuming it's intended that the setter only affects the wrapper, and does not get propagated to the inner env. I'm not sure if this makes any difference in the end, but I'm also not fully immediately convinced which option is better? Is there any use case in which either option is better?

Reply:
That's already the case: #1347

```python
class CustomEnv(gym.Env):
    g = 9.81

class CustomWrapper(gym.Wrapper):
    pass

env = CustomWrapper(CustomEnv())
print(env.g)  # 9.81
env.g = 1.62
print(env.g)  # 1.62
```

The reason why we can't just rely on `__getattr___` for `observation_space`, `action_space`, `reward_range` and `metadata` is that since [`Wrapper` inherits from `Env`](https://github.com/openai/gym/blob/5a4709b158009ba8786c5df2490d8a601bc80ce4/gym/core.py#L212), [those properties will get set](https://github.com/openai/gym/blob/5a4709b158009ba8786c5df2490d8a601bc80ce4/gym/core.py#L32-L39) in every wrapper, and so forwarding will not work properly: for example, if you don't set `observation_space` in your wrapper, because you'd like it to be the same as the wrapped environment, then you'd always get `wrapped_env.observation_space == None`.

Reply:
I don't think mutating `observation_space`/`action_space` in the inner envs is a good idea. There may be cases where you'd like to have access to the space of the original environment (e.g. through `env.unwrapped.action_space`). If the setter was propagated to the inner env, then you'd always have `env.action_space == env.unwrapped.action_space`.